src/activity_item_handling.cpp

```diff
@@ -2759,168 +2767,305 @@ static zone_type_id get_zone_for_act( const tripoint_bub_ms &src_loc, const zone
-            // especially if NPCs have a backlog of moves or there is a lot of them
-            if( !found_route ) {
-                found_route = true;
+    for( auto src_set_iter = src_set.begin(); src_set_iter != src_set.end(); ) {
+        bool skip = false;
+
+        for( const npc &guy : g->all_npcs() ) {
+            if( &guy == &you ) {
                 continue;
             }
-            if( found_one_point ) {
+            if( guy.has_player_activity() && guy.activity.placement == *src_set_iter ) {
+                src_set_iter = src_set.erase( src_set_iter );
+                skip = true;
                 break;
             }
-            for( const item &stack_elem : here.i_at( elem ) ) {
-                if( stack_elem.has_var( "activity_var" ) && stack_elem.get_var( "activity_var", "" ) == you.name ) {
-                    const furn_t &f = here.furn( elem ).obj();
-                    if( !f.has_flag( ter_furn_flag::TFLAG_PLANT ) ) {
-                        src_set.insert( here.get_abs( elem ) );
-                        found_one_point = true;
-                        // only check for a valid path, as that is all that is needed to tidy something up.
-                        if( square_dist( you.pos_bub(), elem ) > 1 ) {
-                            std::vector<tripoint_bub_ms> route = route_adjacent( you, elem );
-                            if( route.empty() ) {
-                                found_route = false;
-                            }
-                        }
-                        break;
-                    }
-                }
-            }
         }
-        if( src_set.empty() && unsorted_spot != tripoint_bub_ms::zero ) {
-            for( const item *inv_elem : you.inv_dump() ) {
-                if( inv_elem->has_var( "activity_var" ) ) {
-                    // we've gone to tidy up all the things lying around, now tidy up the things we picked up.
-                    src_set.insert( here.get_abs( unsorted_spot ) );
-                    break;
-                }
-            }
+        if( skip ) {
+            continue;
+        } else {
+            ++src_set_iter;
         }
-    } else if( act_id == ACT_MULTIPLE_READ ) {
-        // anywhere well lit
-        for( const tripoint_bub_ms &elem : here.points_in_radius( localpos, MAX_VIEW_DISTANCE ) ) {
+    }
+}
+
+void prune_dark_locations( Character &you, std::unordered_set<tripoint_abs_ms> &src_set,
+                           const activity_id &act_id )
+{
+    bool dark_capable = can_do_in_dark( act_id );
+    if( dark_capable ) {
+        return;
+    }
+
+    map &here = get_map();
+    bool src_set_empty = src_set.empty();
+
+    for( auto src_set_iter = src_set.begin(); src_set_iter != src_set.end(); ) {
+        const tripoint_bub_ms set_pt = here.get_bub( *src_set_iter );
+        if( !dark_capable && you.fine_detail_vision_mod( set_pt ) > LIGHT_AMBIENT_DIM ) {
```

Comment:
The `!dark_capable` check ends up being redundant because of the early return. It's *always* true if this has been reached.